api/src/OpenTelemetry/Internal/Common/Types.hs

```diff
@@ -159,22 +166,91 @@ instance ToValue AnyValue where
   toValue = id
 
 
-data ShutdownResult = ShutdownSuccess | ShutdownFailure | ShutdownTimeout
+data ShutdownResult
+  = ShutdownSuccess
+  | ShutdownTimeout
+  | ShutdownError [SomeException]
+
+
+{- | (<>) concatenates the error lists if both arguments are @ShutdownError@, otherwise it returns the worst of the two in this order
+- @ShutdownError@
+- @ShutdownTimeout@
+- @ShutdownSuccess@
+-}
+instance Semigroup ShutdownResult where
+  ShutdownError l <> ShutdownError r = ShutdownError (l <> r)
+  ShutdownError es <> _ = ShutdownError es
+  _ <> ShutdownError es = ShutdownError es
+  ShutdownTimeout <> _ = ShutdownTimeout
+  _ <> ShutdownTimeout = ShutdownTimeout
+  ShutdownSuccess <> ShutdownSuccess = ShutdownSuccess
+
+
+-- | mempty is ShutdownSuccess
+instance Monoid ShutdownResult where
+  mempty = ShutdownSuccess
+
+
+shutdownErrorHandler :: SomeException -> IO ShutdownResult
+shutdownErrorHandler = pure . ShutdownError . pure
+
+
+flushResultToShutdownResult :: FlushResult -> ShutdownResult
+flushResultToShutdownResult FlushSuccess = ShutdownSuccess
+flushResultToShutdownResult FlushTimeout = ShutdownTimeout
+flushResultToShutdownResult (FlushError es) = ShutdownError es
+
+
+takeWorstShutdownResult :: (Foldable t) => t ShutdownResult -> ShutdownResult
+takeWorstShutdownResult = fold
```

Comment:
Developers may be able to use `fold` instead of this function, but I think `takeWorsShutdownResult` better communicates the meaning of the action and `fold` might be less well known